### ISSUE TITLE
Feature w/prototype implementation: Per-task memory profiling

### ISSUE THREAD

**Issue by itamarst** [dask/distributed#2184]
In our Dask Distributed setup, which is currently single machine, I've needed to:

1. Figure out per-task peak memory usage, so we can figure out how many cores to use.
2. Reduce memory usage in ways that improves parallelism, which also means reducing peak memory for tasks that run in parallel.

To do so I have implemented a scheduler plugin that dumps a CSV of per-task memory usage (minimum and peak). Currently it assumes one thread per process, and all-local processes.

In general this has been extremely useful in helping achieve the two goals above, and I'm wondering if this is something worth integrating (in better written & more general form) into Dask Distributed:

```python
class MemoryUsagePlugin(SchedulerPlugin):
    """Record max and min memory usage for a task.

    Assumptions:

    * One task per process: each process has a single thread running a single
      task at a time.

    Limitations:

    * Statistical profiling at 10ms resolution.
    """
    def __init__(self, log_directory: str, cluster: LocalCluster):
        super().__init__()
        f = open(os.path.join(log_directory, "memory_usage.csv"), "w",
                 buffering=1)
        self._csv = csv.writer(f)
        self._csv.writerow(["task_key", "min_memory_mb", "max_memory_mb"])
        self._cluster = cluster
        # Map between worker address ("tcp://127.0.0.1:344") and list of memory
        # usage in MBs. cluster.workers is actually a list of Nanny objects,
        # which are wrappers around the underlying worker process.
        self._memory_per_worker = {worker.process.worker_address: []
                                   for worker in cluster.workers}
        # Schedule memory update every 10ms:
        def update_memory_usage():
            pc = PeriodicCallback(self._io_update_worker_memory_use, 10)
            pc.start()
        cluster.loop.add_callback(update_memory_usage)

    def _io_update_worker_memory_use(self):
        """Record current memory usage for worker processes.

        Runs in Tornado event loop thread.
        """
        for worker in self._cluster.workers:
            if worker.process is None:
                continue
            process = worker.process.process
            if process is None:
                continue
            try:
                proc = psutil.Process(process.pid)
                memory_mb = sum([
                    p.memory_info().rss / (1024 * 1024)
                    for p in [proc] + list(proc.children(recursive=True))
                ])
            except psutil.NoSuchProcess:
                continue
            try:
                self._memory_per_worker[
                    worker.process.worker_address].append(memory_mb)
            except KeyError:
                # Probably has been removed.
                pass

    def _io_clear_worker_memory(self, worker_address: str):
        """Clear memory usage of worker.

        Runs in Tornado event loop thread.
        """
        self._memory_per_worker[worker_address] = []

    def clear_worker_memory(self, worker_address: str):
        """Clear memory usage of worker."""
        self._cluster.loop.add_callback(
            self._io_clear_worker_memory, worker_address)

    def transition(self, key, start, finish, *args, **kwargs):
        """Called by the Scheduler every time a task changes status."""
        # If the task finished, record its memory usage:
        if start == "processing" and finish in ("memory", "erred"):
            worker_url = kwargs["worker"]
            # Get memory usage we've record for the relevant worker, get max
            # and min:
            memory_usage = self._memory_per_worker[kwargs["worker"]][:]
            max_memory_usage = max(memory_usage)
            min_memory_usage = min(memory_usage)
            self._csv.writerow([key, min_memory_usage, max_memory_usage])
            # Wipe memory usage since we're going to start new task now:
            self.clear_worker_memory(kwargs["worker"])
```

**Comment by itamarst** [comment]
Example output:

```csv
task_key,min_memory_mb,max_memory_mb
"('from_sequence-28280073f05e0c7f69f08712ba674712', 0)",165.62890625,165.62890625
"('from_sequence-dc09301fc321e7f820628a8941ff1fdc', 0)",166.0234375,166.0234375
"('product-starmap-starmap-run_filter-5a6d5fd0466442b54b73dae87e899bcd', 22)",175.21484375,1337.515625
"('starmap-run_filter-5a6d5fd0466442b54b73dae87e899bcd', 22)",382.16796875,382.16796875
"('product-starmap-starmap-run_filter-5a6d5fd0466442b54b73dae87e899bcd', 5)",166.03515625,1421.578125
```

**Comment by mrocklin** [comment]
I think that per-task memory use would be *very *valuable.  This is the
sort of thing we could also include in diagnostics.

I'm curious about what to do in the general case where we have multiple
threads actively running tasks at once.

**Comment by itamarst** [comment]
At a minimum, there's the option of saying "do your profiling run with one thread per CPU". This is not ideal, but not awful.

In practice it's quite difficult (or perhaps impossible) to get per-thread information out of Python. Doesn't conceptually make sense given how it works.

Also If people are using threads with objects that are shared in memory it's also hard to allocate which task uses what...

So I'm not sure it's possible to reasonably support multiple threads running at once with the given output.

You _could_ have output which is "for this combination of tasks, we had this peak memory usage". Which is better than nothing. And with sufficiently diverse set of permutations of tasks-combinations might allow estimating task-specific memory usage, maybe.

(Also note the other limitation of prototype, it's LocalCluster-specific. Much more obvious how to fix that though, at least to someone who knows codebase better than I do.)
